torch/csrc/jit/codegen/cuda/multidevice_runtime.h

```diff
@@ -0,0 +1,238 @@
+
+  // Make the given tensor a global input
+  void addFusionInput(TensorView* tv);
+
+  // Interface to call from ir builder which will
+  //  notify any creation of new nodes, so that
+  //  the group placement can be controlled by
+  //  scheduler fully manually.
+  void newStmt(IrBuilderPasskey, Statement* stmt);
+
+ private:
+  auto& currentGroup() {
+    TORCH_INTERNAL_ASSERT(
+        !group_creation_records_.empty(), "call newGroup first.");
+    return group_creation_records_.back();
+  }
+
+  std::unique_ptr<SegmentedGroup> buildGroup(const GroupRecord& group_record);
+
+ private:
+  //! Keeps track of user decided group segmentation
+  //!  in scheduling time.
+  std::vector<GroupRecord> group_creation_records_;
+
+  //! Keeps track of currently available tensorviews to
+  //!  avoid re-computation.
+  std::unordered_map<TensorView*, GroupRecord*> context_tensor_map_;
+
+  //! Original fusion this builder will be
+  //!  spliting. Using an owning pointer to
+  //!  avoid redundant copying.
+  std::unique_ptr<Fusion> original_fusion_ = nullptr;
+
+  //! Running counter to generate unique group id.
+  int running_group_counter_ = 0;
+
+  //! Keep track of validity of the current builder.
+  //!  each builder can only produce multi-group fusion once.
+  bool valid_ = true;
+};
+
+//! Runtime to support running multi-group fusion on
+//!  multiple devices.
+class TORCH_CUDA_CU_API MultiDeviceRuntime {
+  using CompiledKernelPtr = std::unique_ptr<FusionExecutor>;
+
+ public:
+  explicit MultiDeviceRuntime(
+      std::unique_ptr<MultiGroupFusion> multi_group_fusion)
+      : multi_group_fusion_(std::move(multi_group_fusion)) {}
+
+  // Run kernels with the given global inputs, compile if needed.
+  std::vector<at::Tensor> runWithInput(std::vector<IValue> inputs);
+
+  // Interface to querry underlying fusion.
+  auto multiGroupFusion() const {
+    return multi_group_fusion_.get();
+  }
+
+  // Interface short-cut to querry the flattened fusion
+  //  containing all the expressions in all groups.
+  auto flattenedFusion() const {
+    return multi_group_fusion_->completeFusion();
+  }
+
+ private:
+  // Get list of global inputs to the complete fusion.
+  const auto& globalInputs() const {
+    return multi_group_fusion_->completeFusion()->inputs();
+  }
+
+  // Get list of global outputs from the complete fusion.
+  const auto& globalOutputs() const {
+    return multi_group_fusion_->completeFusion()->outputs();
+  }
+
+  // Generate and compile cuda kernel corresponding to
+  //  the given segmented group.
+  CompiledKernelPtr compileGroup(
+      SegmentedGroup* group,
+      std::vector<IValue> group_input);
+
+  // Retrieve all inputs corresponding to the given group from
+  //  the current context.
+  std::vector<IValue> getGroupIValueInputs(SegmentedGroup*);
+
+  // Retrieve computed values from the current context,
+  //  throws an error if the given val hasn't been computed.
+  inline IValue getIValueFromFusionVal(Val* val);
```

Comment:
I don't like this direct binding of IValue's into this container. We should separate out runtime from compilation, and this feels like it mixes the two.